pkg/operator/controllers/azurensg/subnetnsg.go

```diff
@@ -0,0 +1,224 @@
+	if err != nil {
+		return err
+	}
+
+	vnetResource, err := azure.ParseResourceID(instance.Spec.VnetID)
+	if err != nil {
+		return err
+	}
+
+	workerSubnets, err := r.workerSubnets(ctx, r.maocli)
+	if err != nil {
+		return err
+	}
+
+	subnetsClient := network.NewSubnetsClient(&azEnv, resource.SubscriptionID, authorizer)
+	switch admin.ArchitectureVersion(instance.Spec.ArchitectureVersion) {
+	case admin.ArchitectureVersionV1: // int == 0
+		//master subnet
+		v, err := r.isSubnetNSGValid(ctx, masterSubnet, vnetResource, resource, masterResourceGroup, subnetsClient, instance.Spec.InfraID+"-controlplane-nsg")
+		if err != nil {
+			return err
+		}
+
+		if !v {
+			err = r.fixSubnetNSG(ctx, masterSubnet, vnetResource, resource, masterResourceGroup, subnetsClient, instance.Spec.InfraID+"-controlplane-nsg", instance.Spec.InfraID)
+			if err != nil {
+				return err
+			}
+		}
+		//worker subnets
+		for _, ws := range workerSubnets {
+			v, err := r.isSubnetNSGValid(ctx, ws, vnetResource, resource, masterResourceGroup, subnetsClient, instance.Spec.InfraID+"-node-nsg")
+			if err != nil {
+				return err
+			}
+			if !v {
+				err = r.fixSubnetNSG(ctx, ws, vnetResource, resource, masterResourceGroup, subnetsClient, instance.Spec.InfraID+"-node-nsg", instance.Spec.InfraID)
+				if err != nil {
+					return err
+				}
+			}
+		}
+	case admin.ArchitectureVersionV2: // int == 1
+		//master subnet
+		v, err := r.isSubnetNSGValid(ctx, masterSubnet, vnetResource, resource, masterResourceGroup, subnetsClient, instance.Spec.InfraID+"-nsg")
+		if err != nil {
+			return err
+		}
+
+		if !v {
+			err = r.fixSubnetNSG(ctx, masterSubnet, vnetResource, resource, masterResourceGroup, subnetsClient, instance.Spec.InfraID+"-nsg", instance.Spec.InfraID)
+			if err != nil {
+				return err
+			}
+		}
+		//worker subnets
+		for _, ws := range workerSubnets {
+			v, err := r.isSubnetNSGValid(ctx, ws, vnetResource, resource, masterResourceGroup, subnetsClient, instance.Spec.InfraID+"-nsg")
+			if err != nil {
+				return err
+			}
+			if !v {
+				err = r.fixSubnetNSG(ctx, ws, vnetResource, resource, masterResourceGroup, subnetsClient, instance.Spec.InfraID+"-nsg", instance.Spec.InfraID)
+				if err != nil {
+					return err
+				}
+			}
+		}
+	default:
+		r.log.Errorf("Unknown value of Architecture Version %d", instance.Spec.ArchitectureVersion)
+	}
+	return nil
+}
+
+func (r *AzureNSGReconciler) fixSubnetNSG(ctx context.Context, subnet string, vnet, resource azure.Resource, masterResourceGroup string, subnetsClient network.SubnetsClient, expectedName, infraID string) error {
+	fullSubnet, err := subnetsClient.Get(ctx, vnet.ResourceGroup, vnet.ResourceName, subnet, "")
+	if err != nil {
+		return err
+	}
+	fullSubnet.NetworkSecurityGroup = &mgmtnetwork.SecurityGroup{ID: to.StringPtr("/subscriptions/" + resource.SubscriptionID + "/resourceGroups/" + masterResourceGroup + "/providers/Microsoft.Network/networkSecurityGroups/" + infraID + "-nsg")}
```

Comment:
Is the name of the nsg  correct here ? should not it be expectedName instead of infraId+nsg ?

Reply:
(to cover arch v1 and v2 cases)

Reply:
Thanks,
I fixed it in code and changed that in the unit test which now doesn't assume "-nsg" suffix.